quiche/src/recovery/gcongestion/recovery.rs

```diff
@@ -0,0 +1,1002 @@
+    }
+
+    #[cfg(test)]
+    fn bytes_in_flight(&self) -> usize {
+        self.bytes_in_flight
+    }
+
+    #[cfg(test)]
+    fn pacing_rate(&self) -> u64 {
+        self.pacer
+            .pacing_rate(self.bytes_in_flight, &self.rtt_stats)
+            .to_bytes_per_period(Duration::from_secs(1))
+    }
+
+    #[cfg(test)]
+    fn pto_count(&self) -> u32 {
+        self.pto_count
+    }
+
+    #[cfg(test)]
+    fn pkt_thresh(&self) -> u64 {
+        self.pkt_thresh
+    }
+
+    #[cfg(test)]
+    fn lost_spurious_count(&self) -> usize {
+        self.lost_spurious_count
+    }
+
+    #[cfg(test)]
+    fn detect_lost_packets_for_test(
+        &mut self, epoch: packet::Epoch, now: Instant,
+    ) -> (usize, usize) {
+        let ret = self.detect_and_remove_lost_packets(epoch, now);
+        self.epochs[epoch].drain_acked_and_lost_packets();
+        ret
+    }
+
+    #[cfg(test)]
+    fn app_limited(&self) -> bool {
+        self.pacer.is_app_limited(self.bytes_in_flight)
+    }
+
+    fn update_app_limited(&mut self, _v: bool) {
+        // TODO
+    }
+
+    fn delivery_rate_update_app_limited(&mut self, _v: bool) {
+        // TODO
+    }
+
+    fn update_max_ack_delay(&mut self, max_ack_delay: Duration) {
+        self.rtt_stats.max_ack_delay = max_ack_delay;
+    }
+
+    fn get_next_release_time(&self) -> ReleaseDecision {
+        self.pacer.get_next_release_time()
+    }
+
+    fn use_get_next_release_time(&self) -> bool {
+        true
+    }
+
+    #[cfg(feature = "qlog")]
+    fn maybe_qlog(&mut self) -> Option<EventData> {
+        let qlog_metrics = QlogMetrics {
+            min_rtt: self.rtt_stats.min_rtt().unwrap_or(Duration::MAX),
+            smoothed_rtt: self.rtt(),
+            latest_rtt: self.rtt_stats.latest_rtt(),
+            rttvar: self.rtt_stats.rttvar(),
+            cwnd: self.cwnd() as u64,
+            bytes_in_flight: self.bytes_in_flight as u64,
+            ssthresh: self.pacer.ssthresh(),
+            pacing_rate: self.delivery_rate(),
+        };
+
+        self.qlog_metrics.maybe_update(qlog_metrics)
+    }
+
+    fn send_quantum(&self) -> usize {
+        let pacing_rate = self
+            .pacer
+            .pacing_rate(self.bytes_in_flight, &self.rtt_stats);
+
+        let floor = if pacing_rate < Bandwidth::from_kbits_per_second(1200) {
+            self.max_datagram_size
+        } else {
+            2 * self.max_datagram_size
+        };
+
+        pacing_rate
+            .to_bytes_per_period(ReleaseDecision::EQUAL_THRESHOLD)
+            .min(64 * 1024)
+            .max(floor as u64) as usize
+    }
+}
+
+impl std::fmt::Debug for GRecovery {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "timer={:?} ", self.loss_detection_timer())?;
+        write!(f, "rtt_stats={:?} ", self.rtt_stats)?;
+        write!(f, "bytes_in_flight={} ", self.bytes_in_flight)?;
+        write!(f, "{:?} ", self.pacer)?;
+        Ok(())
+    }
+}
```

Comment:
This custom format makes me wonder if its being used for logging/metrics/tests. Do we need to preserve this for now?